code/datums/components/pellet_cloud.dm

```diff
@@ -0,0 +1,249 @@
+	*	-Directed: This means you're shooting multiple pellets, like buckshot. If an ammo casing is defined as having multiple pellets, it will automatically create a pellet cloud
+	*		and call COMSIG_PELLET_CLOUD_INIT (see [/obj/item/ammo_casing/proc/fire_casing]). Thus, the only projectiles fired will be the ones fired here.
+	*		The magnitude var controls how many pellets are created.
+	*	-Circular: This results in a big spray of shrapnel flying all around the detonation point when the grenade fires COMSIG_GRENADE_PRIME or landmine triggers COMSIG_MINE_TRIGGERED.
+	*		The magnitude var controls how big the detonation radius is (the bigger the magnitude, the more shrapnel is created). Grenades can be covered with bodies to reduce shrapnel output.
+	*
+	* Once all of the fired projectiles either hit a target or disappear due to ranging out/whatever else, we resolve the list of all the things we hit and print aggregate messages so we get
+	* one "You're hit by 6 buckshot pellets" vs 6x "You're hit by the buckshot blah blah" messages.
+	*
+	* Note that this is how all guns handle shooting ammo casings with multiple pellets, in case such a thing comes up.
+*/
+
+/datum/component/pellet_cloud
+	var/projectile_type /// What's the projectile path of the shrapnel we're shooting?
+
+	var/num_pellets /// How many shrapnel projectiles are we responsible for tracking? May be reduced for grenades if someone dives on top of it. Defined by ammo casing for casings, derived from magnitude otherwise
+	var/radius = 4 /// For grenades/landmines, how big is the radius of turfs we're targeting? Note this does not effect the projectiles range, only how many we generate
+
+	var/list/pellets = list() /// The list of pellets we're responsible for tracking, once these are all accounted for, we finalize.
+	var/list/targets_hit = list() /// An associated list with the atom hit as the key and how many pellets they've eaten for the value, for printing aggregate messages
+	var/list/bodies /// For grenades, any /mob/living's the grenade is moved onto, see [/datum/component/pellet_cloud/proc/handle_martyrs()]
+	var/list/purple_hearts /// For grenades, tracking people who die covering a grenade for achievement purposes, see [/datum/component/pellet_cloud/proc/handle_martyrs()]
+
+	var/terminated /// how many pellets ranged out without hitting anything
+	var/hits /// how many pellets impacted something
+
+	var/not_done_yet /// for grenades and landmines (and I guess caseless ammo casings actually, goddammit) where the parent deletes itself before all the pellets are accounted for
+
+	var/mob/living/shooter /// for if we're an ammo casing being fired
+
+/datum/component/pellet_cloud/Initialize(projectile_type=/obj/item/shrapnel, magnitude=5)
+	if(!isammocasing(parent) && !isgrenade(parent) && !islandmine(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	if(magnitude < 1)
+		CRASH("Invalid magnitude [magnitude] < 1 on pellet_cloud, parent: [parent]")
+		qdel(src, TRUE)
+		return
```

Comment:
`CRASH` acts like a `return`
The code under it won't execute.
Use `stack_trace` instead.